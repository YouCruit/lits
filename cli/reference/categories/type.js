module.exports = {
  'type-of': {
    name: `type-of`,
    category: `Type`,
    linkName: `type-of`,
    returns: {
      type: `type`,
    },
    arguments: [
      {
        name: `value`,
        type: `Any`,
      },
    ],
    description: `Returns the type of \`value\`.`,
    examples: [
      `(type-of -1.1)`,
      `(type-of -1)`,
      `(type-of 0)`,
      `(type-of 1)`,
      `(type-of 1.1)`,
      `(type-of true)`,
      `(type-of false)`,
      `(type-of (/ 0 0))`,
      `(type-of (/ 1 0))`,
      `(type-of (/ -1 0))`,
      `(type-of "")`,
      `(type-of :Foo)`,
      `(type-of [])`,
      `(type-of [1 2 3])`,
      `(type-of {})`,
      `(type-of { :foo :bar })`,
      `(type-of #"^foo")`,
      `(type-of #(identity %1))`,
      `(type-of nil)`,
    ],
  },
  'type-or': {
    name: `type-or`,
    category: `Type`,
    linkName: `type-or`,
    returns: {
      type: `type`,
    },
    arguments: [
      {
        name: `types`,
        type: `type[]`,
        description: `one or many`,
      },
    ],
    description: `Returns the disjuntion of the \`types\`.`,
    examples: [
      `(type-or ::float ::positive-infinity ::negative-infinity ::nan)`,
      `(type-or ::float ::true ::false)`,
      `(type-or ::positive-integer ::negative-integer ::zero)`,
    ],
  },
  'type-and': {
    name: `type-and`,
    category: `Type`,
    linkName: `type-and`,
    returns: {
      type: `type`,
    },
    arguments: [
      {
        name: `types`,
        type: `type[]`,
        description: `one or many`,
      },
    ],
    description: `Return the conjuction of the \`types\`.`,
    examples: [
      `(type-and ::float ::integer)`,
      `(type-and ::float ::string)`,
      `(type-and ::non-positive-integer ::non-negative-float)`,
    ],
  },
  'type-exclude': {
    name: `type-exclude`,
    category: `Type`,
    linkName: `type-exclude`,
    returns: {
      type: `type`,
    },
    arguments: [
      {
        name: `types`,
        type: `type[]`,
        description: `one or many`,
      },
    ],
    description: `Subtract types from the first parameter.`,
    examples: [
      `(type-exclude ::float ::integer)`,
      `(type-exclude ::float ::string)`,
      `(type-exclude ::float ::non-negative-float)`,
    ],
  },
  'type-is?': {
    name: `type-is?`,
    category: `Type`,
    linkName: `type-is_question`,
    returns: {
      type: `boolean`,
    },
    arguments: [
      {
        name: `type1`,
        type: `type`,
      },
      {
        name: `type2`,
        type: `type`,
      },
    ],
    description: `Check if \`type1\` is of type \`type2\`.`,
    examples: [
      `(type-is? ::float ::integer)`,
      `(type-is? ::integer ::float)`,
      `(type-is? ::empty-string ::string)`,
      `(type-is? ::empty-string ::falsy)`,
      `(type-is? ::empty-string ::truthy)`,
    ],
  },
  'type-equals?': {
    name: `type-equals?`,
    category: `Type`,
    linkName: `type-equals_question`,
    returns: {
      type: `boolean`,
    },
    arguments: [
      {
        name: `type1`,
        type: `type`,
      },
      {
        name: `type2`,
        type: `type`,
      },
    ],
    description: `Check if \`type1\` is equal to \`type2\`.`,
    examples: [
      `(type-equals? ::integer ::float)`,
      `(type-equals? ::empty-string ::string)`,
      `(type-equals? ::empty-string (type-and ::falsy ::boolean))`,
    ],
  },
  'type-intersects?': {
    name: `type-intersects?`,
    category: `Type`,
    linkName: `type-intersects_question`,
    returns: {
      type: `boolean`,
    },
    arguments: [
      {
        name: `type1`,
        type: `type`,
      },
      {
        name: `type2`,
        type: `type`,
      },
    ],
    description: `Check if \`type1\` intersects \`type2\`.`,
    examples: [
      `(type-intersects? ::float ::integer)`,
      `(type-intersects? ::integer ::float)`,
      `(type-intersects? ::empty-string ::string)`,
      `(type-intersects? ::empty-string ::falsy)`,
      `(type-intersects? ::empty-string ::truthy)`,
    ],
  },
}
