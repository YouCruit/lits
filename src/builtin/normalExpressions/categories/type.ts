import { DataType } from '../../../Lits/Lits'
import { any, assertNumberOfParams, dataType } from '../../../utils/assertion'
import { BuiltinNormalExpressions } from '../../interface'

export const typeNormalExpression: BuiltinNormalExpressions = {
  'type-of': {
    evaluate: ([value], debugInfo): DataType => {
      any.assert(value, debugInfo)
      return DataType.of(value)
    },
    validate: node => assertNumberOfParams(1, node),
  },
  'type-or': {
    evaluate: (params, debugInfo): DataType => {
      params.forEach(param => dataType.assert(param, debugInfo))
      return DataType.or(...(params as DataType[]))
    },
    validate: () => undefined,
  },
  'type-and': {
    evaluate: (params, debugInfo): DataType => {
      params.forEach(param => dataType.assert(param, debugInfo))
      return DataType.and(...(params as DataType[]))
    },
    validate: () => undefined,
  },
  'type-exclude': {
    evaluate: (params, debugInfo): DataType => {
      params.forEach(param => dataType.assert(param, debugInfo))
      const first = dataType.as(params[0])
      return DataType.exclude(first, ...(params.slice(1) as DataType[]))
    },
    validate: node => assertNumberOfParams({ min: 1 }, node),
  },
  'type-is?': {
    evaluate: ([first, second], debugInfo): boolean => {
      dataType.assert(first, debugInfo)
      dataType.assert(second, debugInfo)
      return DataType.is(first, second)
    },
    validate: node => assertNumberOfParams(2, node),
  },
  'type-equals?': {
    evaluate: ([first, second], debugInfo): boolean => {
      dataType.assert(first, debugInfo)
      dataType.assert(second, debugInfo)
      return DataType.equals(first, second)
    },
    validate: node => assertNumberOfParams(2, node),
  },
  'type-intersects?': {
    evaluate: ([first, second], debugInfo): boolean => {
      dataType.assert(first, debugInfo)
      dataType.assert(second, debugInfo)
      return DataType.intersects(first, second)
    },
    validate: node => assertNumberOfParams(2, node),
  },
}
