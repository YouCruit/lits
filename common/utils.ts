export function throttle<T extends (...args: any[]) => void>(func: T) {
  let openForBusiness = true
  return function (this: any, ...args: Parameters<T>) {
    if (openForBusiness) {
      requestAnimationFrame(() => openForBusiness = true)
      openForBusiness = false
      func.apply(this, args)
    }
  }
}
