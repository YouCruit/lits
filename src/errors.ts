/* istanbul ignore file */

import { AstNode, NodeType } from './parser/interface'
import { Token } from './tokenizer/interface'

export class RecurSignal extends Error {
  public params: unknown[]
  constructor(params: unknown[]) {
    super(`recur, params: ${params}`)
    Object.setPrototypeOf(this, RecurSignal.prototype)
    this.name = `RecurSignal`
    this.params = params
  }
}

export class UserDefinedError extends Error {
  constructor(message: string) {
    super(message)
    Object.setPrototypeOf(this, UserDefinedError.prototype)
    this.name = `UserDefinedError`
  }
}

export class UnexpectedTokenError extends Error {
  constructor(expectedToken: string, actualToken: Token) {
    super(`Expected a "${expectedToken}" token, got Token[${actualToken.type}:"${actualToken.value}"]`)
    Object.setPrototypeOf(this, UnexpectedTokenError.prototype)
    this.name = `UnexpectedTokenError`
  }
}

export class UnexpectedNodeTypeError extends Error {
  constructor(expectedNodeType: NodeType | `ExpressionNode`, actualNode: AstNode | undefined) {
    super(`Expected a ${expectedNodeType} node, got ${actualNode ? `a ${actualNode.type} node` : `undefined`}`)
    Object.setPrototypeOf(this, UnexpectedNodeTypeError.prototype)
    this.name = `UnexpectedNodeTypeError`
  }
}
