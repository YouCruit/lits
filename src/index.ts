import { evaluate } from './evaluator'
import { Context, VariableScope } from './evaluator/interface'
import { parse } from './parser'
import { Ast } from './parser/interface'
import { tokenize } from './tokenizer'
import { Token } from './tokenizer/interface'

export { evaluate } from './evaluator'
export { Context } from './evaluator/interface'
export { parse } from './parser'
export { Ast } from './parser/interface'
export { tokenize } from './tokenizer'
export { Token } from './tokenizer/interface'

export function lispish(
  program: string,
  globalVariables: VariableScope = {},
  topScope: Context = { variables: {} },
): unknown {
  const tokens: Token[] = tokenize(program)
  const ast: Ast = parse(tokens)
  const result = evaluate(ast, globalVariables, topScope)
  return result
}

export { normalExpressionKeys, specialExpressionKeys } from './builtin'
export { reservedNames } from './reservedNames'
