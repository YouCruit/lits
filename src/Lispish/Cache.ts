import { asNotUndefined, toNonNegativeInteger } from '../utils'

type CacheEntry<T> = {
  key: string
  value: T
  nextEntry: CacheEntry<T> | undefined
}

export class Cache<T> {
  private cache: Record<string, CacheEntry<T>> = {}
  private firstEntry: CacheEntry<T> | undefined = undefined
  private lastEntry: CacheEntry<T> | undefined = undefined
  private _size = 0
  private maxSize: number
  constructor(maxSize: number) {
    this.maxSize = toNonNegativeInteger(maxSize)
    if (this.maxSize < 1) {
      throw Error(`1 is the minimum maxSize, got ${maxSize}`)
    }
  }

  public get size(): number {
    return this._size
  }

  public get(key: string): T | undefined {
    return this.cache[key]?.value
  }

  public clear(): void {
    this.cache = {}
    this.firstEntry = undefined
    this.lastEntry = undefined
    this._size = 0
  }

  public has(key: string): boolean {
    return !!this.cache[key]
  }

  public set(key: string, value: T): void {
    if (this.has(key)) {
      throw Error(`AstCache - key already present: ${key}`)
    }
    const newEntry: CacheEntry<T> = { value, nextEntry: undefined, key }

    this.cache[key] = newEntry
    this._size += 1

    if (this.lastEntry) {
      this.lastEntry.nextEntry = newEntry
    }
    this.lastEntry = newEntry

    if (!this.firstEntry) {
      this.firstEntry = this.lastEntry
    }

    while (this.size > this.maxSize) {
      this.dropFirstEntry()
    }
  }

  private dropFirstEntry(): void {
    const firstEntry = asNotUndefined(this.firstEntry)
    delete this.cache[firstEntry.key]
    this._size -= 1
    this.firstEntry = firstEntry.nextEntry
  }
}
