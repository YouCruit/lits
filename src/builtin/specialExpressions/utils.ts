import { AstNode, ParseArgument } from '../../parser/interface'
import { Token } from '../../tokenizer/interface'
import { asNotUndefined } from '../../utils'

export type FunctionArguments = {
  mandatoryArguments: string[]
  optionalArguments: Array<{
    name: string
    defaultValue?: AstNode
  }>
  restArgument?: string
}
export function parseFunctionArguments(
  tokens: Token[],
  position: number,
  parseArgument: ParseArgument,
): [number, FunctionArguments] {
  const args: FunctionArguments = {
    mandatoryArguments: [],
    optionalArguments: [],
  }
  const argNames: Record<string, true> = {}
  let state: 'mandatory' | 'optional' | 'rest' = 'mandatory'
  let token = asNotUndefined(tokens[position])
  while (!(token.type === 'paren' && token.value === ')')) {
    const [newPosition, node] = parseArgument(tokens, position)
    position = newPosition
    token = asNotUndefined(tokens[position])

    if (node.type === 'Modifier') {
      switch (node.value) {
        case '&optional':
          if (state === 'rest') {
            throw Error('&optional cannot appear after &rest')
          }
          if (state === 'optional') {
            throw Error('&optional can only appear once')
          }
          state = 'optional'
          break
        case '&rest':
          if (state === 'rest') {
            throw Error('&rest can only appear once')
          }
          if (state === 'optional' && args.optionalArguments.length === 0) {
            throw Error('No optional arguments where spcified')
          }
          state = 'rest'
          break
      }
    } else {
      if (argNames[node.name]) {
        throw Error(`Duplicate argument "${node.name}"`)
      } else {
        argNames[node.name] = true
      }

      if (Object.getOwnPropertyDescriptor(node, 'defaultValue')) {
        if (state !== 'optional') {
          throw Error('Cannot specify default value if not an optional argument')
        }
        args.optionalArguments.push({
          name: node.name,
          defaultValue: node.defaultValue,
        })
      } else {
        switch (state) {
          case 'mandatory':
            args.mandatoryArguments.push(node.name)
            break
          case 'optional':
            args.optionalArguments.push({
              name: node.name,
              defaultValue: undefined,
            })
            break
          case 'rest':
            if (args.restArgument) {
              throw Error('Can only specify one rest argument')
            }
            args.restArgument = node.name
            break
        }
      }
    }
  }

  if (state === 'rest' && !Object.getOwnPropertyDescriptor(args, 'restArgument')) {
    throw Error('Missing rest argument name')
  }
  if (state === 'optional' && args.optionalArguments.length === 0) {
    throw Error('No optional arguments where spcified')
  }

  position += 1
  return [position, args]
}
