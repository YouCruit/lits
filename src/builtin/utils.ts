import { UnexpectedTokenError } from '../errors'
import { Context } from '../evaluator/interface'
import { AstNode, BindingNode, ParseArgument, ParseBindings } from '../parser/interface'
import { reservedNamesRecord } from '../reservedNames'
import { Token } from '../tokenizer/interface'
import { asNotUndefined } from '../utils'
import { Builtin } from './interface'

export type FunctionArguments = {
  mandatoryArguments: string[]
  optionalArguments: Array<{
    name: string
    defaultValue?: AstNode
  }>
  restArgument?: string
  bindings: BindingNode[]
}

export function parseFunctionArguments(
  tokens: Token[],
  position: number,
  parseArgument: ParseArgument,
  parseBindings: ParseBindings,
): [number, FunctionArguments] {
  let bindings: BindingNode[] = []
  let restArgument: string | undefined = undefined
  const mandatoryArguments: string[] = []
  const optionalArguments: Array<{
    name: string
    defaultValue?: AstNode
  }> = []
  const argNames: Record<string, true> = {}
  let state: `mandatory` | `optional` | `rest` | `bind` = `mandatory`
  let token = asNotUndefined(tokens[position])
  if (!(token.type === `paren` && token.value === `[`)) {
    throw new UnexpectedTokenError(`[`, token)
  }

  position += 1
  token = asNotUndefined(tokens[position])
  while (!(token.type === `paren` && token.value === `]`)) {
    if (state === `bind`) {
      ;[position, bindings] = parseBindings(tokens, position)
      break
    } else {
      const [newPosition, node] = parseArgument(tokens, position)
      position = newPosition
      token = asNotUndefined(tokens[position])

      if (node.type === `Modifier`) {
        switch (node.value) {
          case `&opt`:
            if (state === `rest`) {
              throw Error(`&opt cannot appear after &rest`)
            }
            if (state === `optional`) {
              throw Error(`&opt can only appear once`)
            }
            state = `optional`
            break
          case `&rest`:
            if (state === `rest`) {
              throw Error(`&rest can only appear once`)
            }
            if (state === `optional` && optionalArguments.length === 0) {
              throw Error(`No optional arguments where spcified`)
            }
            state = `rest`
            break
          case `&bind`:
            if (state === `optional` && optionalArguments.length === 0) {
              throw Error(`No optional arguments where spcified`)
            }
            if (state === `rest` && !restArgument) {
              throw Error(`No rest argument was spcified`)
            }
            state = `bind`
            break
        }
      } else {
        if (argNames[node.name]) {
          throw Error(`Duplicate argument "${node.name}"`)
        } else {
          argNames[node.name] = true
        }

        if (Object.getOwnPropertyDescriptor(node, `defaultValue`)) {
          if (state !== `optional`) {
            throw Error(`Cannot specify default value if not an optional argument`)
          }
          optionalArguments.push({
            name: node.name,
            defaultValue: node.defaultValue,
          })
        } else {
          switch (state) {
            case `mandatory`:
              mandatoryArguments.push(node.name)
              break
            case `optional`:
              optionalArguments.push({
                name: node.name,
                defaultValue: undefined,
              })
              break
            case `rest`:
              if (restArgument !== undefined) {
                throw Error(`Can only specify one rest argument`)
              }
              restArgument = node.name
              break
          }
        }
      }
    }
  }

  if (state === `rest` && restArgument === undefined) {
    throw Error(`Missing rest argument name`)
  }
  if (state === `optional` && optionalArguments.length === 0) {
    throw Error(`No optional arguments where spcified`)
  }

  position += 1

  const args: FunctionArguments = {
    mandatoryArguments,
    optionalArguments,
    restArgument,
    bindings,
  }

  return [position, args]
}

export function assertNameNotDefined<T>(name: T, contextStack: Context[], builtin: Builtin): asserts name is T {
  if (typeof name !== `string`) {
    return
  }
  if (builtin.specialExpressions[name]) {
    throw Error(`Cannot define variable ${name}, it's a special expression`)
  }

  if (builtin.normalExpressions[name]) {
    throw Error(`Cannot define variable ${name}, it's a builtin function`)
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  if ((reservedNamesRecord as any)[name]) {
    throw Error(`Cannot define variable ${name}, it's a reserved name`)
  }

  const globalContext = asNotUndefined(contextStack[contextStack.length - 2])

  if (globalContext[name]) {
    throw Error(`Name already defined "${name}"`)
  }
}
