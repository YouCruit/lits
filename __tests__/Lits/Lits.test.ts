import type { Ast } from '../../src'
import { Lits } from '../../src'
import { UndefinedSymbolError } from '../../src/errors'
import { Cache } from '../../src/Lits/Cache'
import type { LazyValue } from '../../src/Lits/Lits'
import { AstNodeType } from '../../src/constants/constants'
import { TokenType } from '../../src/constants/constants'
import { assertLitsFunction } from '../../src/utils/functionAsserter'

describe(`TEST`, () => {
  let lits: Lits
  beforeEach(() => {
    lits = new Lits({ debug: true, astCacheSize: 0 })
  })
  test(`without params`, () => {
    const fn = lits.run(`#(+ %1 %2)`)
    assertLitsFunction(fn)
    expect(lits.apply(fn, [2, 3])).toBe(5)
  })
  test(`with empty params`, () => {
    const fn = lits.run(`#(+ %1 %2)`)
    assertLitsFunction(fn)
    expect(lits.apply(fn, [2, 3], {})).toBe(5)
  })

  test(`with params`, () => {
    const fn = lits.run(`#(+ %1 %2 x)`)
    assertLitsFunction(fn)
    expect(lits.apply(fn, [2, 3], { contexts: [{ x: { value: 1 } }] })).toBe(6)
  })
})

describe(`Lazy host values as function`, () => {
  test(`that it works`, () => {
    const lits = new Lits()
    const lazyHostValues: Record<string, LazyValue> = {
      x: {
        read: () => 42,
      },
      foo: {
        read: () => ({
          Î»: true,
          t: 301,
          o: [
            {
              as: {
                mandatoryArguments: [],
              },
              a: 0,
              b: [
                {
                  t: 201,
                  v: 42,
                },
              ],
              f: {},
            },
          ],
        }),
      },
    }

    // expect(lits.run(`x`, { lazyValues: lazyHostValues })).toBe(42)
    expect(lits.run(`(foo)`, { lazyValues: lazyHostValues })).toBe(42)
    expect(lits.run(`z`, { lazyValues: { z: { read: () => 12 } } })).toBe(12)
  })
})

describe(`runtime info`, () => {
  test(`getRuntimeInfo().`, () => {
    const lits = new Lits()
    expect(lits.getRuntimeInfo()).toMatchSnapshot()
  })
  test(`getRuntimeInfo() with ast cache > 0`, () => {
    const lits = new Lits({ astCacheSize: 10 })
    expect(lits.getRuntimeInfo()).toMatchSnapshot()
  })
  test(`getRuntimeInfo() with ast cache = 0`, () => {
    const lits = new Lits({ astCacheSize: 0 })
    expect(lits.getRuntimeInfo()).toMatchSnapshot()
  })
})

describe(`context`, () => {
  let lits: Lits
  beforeEach(() => {
    lits = new Lits({ debug: true })
  })
  test(`a function.`, () => {
    lits = new Lits({ astCacheSize: 10 })
    const contexts = [lits.context(`(defn tripple [x] (* x 3))`)]
    expect(lits.run(`(tripple 10)`, { contexts })).toBe(30)
    expect(lits.run(`(tripple 10)`, { contexts })).toBe(30)
  })

  test(`a function - no cache`, () => {
    lits = new Lits({ debug: true })
    const contexts = [lits.context(`(defn tripple [x] (* x 3))`, {})]
    expect(lits.run(`(tripple 10)`, { contexts })).toBe(30)
    expect(lits.run(`(tripple 10)`, { contexts })).toBe(30)
  })

  test(`a function - initial cache`, () => {
    const initialCache: Record<string, Ast> = {
      '(pow 2 4)': {
        b: [
          {
            t: AstNodeType.NormalExpression,
            n: `pow`,
            p: [
              {
                t: AstNodeType.Number,
                v: 2,
                tkn: {
                  t: TokenType.Number,
                  v: `2`,
                },
              },
              {
                t: AstNodeType.Number,
                v: 4,
                tkn: {
                  t: TokenType.Number,
                  v: `4`,
                },
              },
            ],
            tkn: {
              t: TokenType.Name,
              v: `pow`,
            },
          },
        ],
      },
    }
    lits = new Lits({ astCacheSize: 10, initialCache })
    expect(lits.run(`(pow 2 2)`)).toBe(4)
    expect(lits.run(`(pow 2 4)`)).toBe(16)
  })

  test(`a variable.`, () => {
    const contexts = [lits.context(`(def magicNumber 42)`)]
    expect(lits.run(`magicNumber`, { contexts })).toBe(42)
  })

  test(`a variable - again.`, () => {
    const contexts = [
      lits.context(`
    (defn zip? [string] (boolean (match (regexp "^\\d{5}$") string)))
    (defn isoDateString? [string] (boolean (match (regexp "^\\d{4}-\\d{2}-\\d{2}$") string)))
    (def NAME_LENGTH 100)
    `),
    ]
    expect(lits.run(`NAME_LENGTH`, { contexts })).toBe(100)
  })

  test(`change imported variable`, () => {
    const contexts = [lits.context(`(def magicNumber 42)`)]
    expect(lits.run(`magicNumber`, { contexts })).toBe(42)
  })

  test(`a function with a built in normal expression name`, () => {
    expect(() => lits.context(`(defn inc (x) (+ x 1))`)).toThrow()
    expect(() => lits.context(`(defn inc (x) (+ x 1))`, { contexts: [{}] })).toThrow()
    expect(() => lits.context(`(defn inc (x) (+ x 1))`, { values: {} })).toThrow()
  })

  test(`a function with a built in special expression name`, () => {
    expect(() => lits.context(`(defn and (x y) (* x y))`)).toThrow()
  })

  test(`a variable twice`, () => {
    const contexts = [lits.context(`(def magicNumber 42) (defn getMagic [] 42)`)]
    lits.context(`(def magicNumber 42) (defn getMagic [] 42)`, { contexts })
  })

  test(`more than one`, () => {
    const contexts = [lits.context(`(defn tripple [x] (* x 3))`), lits.context(`(def magicNumber 42)`)]
    expect(lits.run(`(tripple magicNumber)`, { contexts })).toBe(126)
  })
})

function ast(n: number): Ast {
  return {
    b: [
      {
        t: AstNodeType.Number,
        v: n,
      },
    ],
  }
}

describe(`Cache`, () => {
  test(`cannot set same key twice`, () => {
    const cache = new Cache(10)
    cache.set(`a`, ast(1))
    expect(() => cache.set(`a`, ast(2))).toThrow()
  })

  test(`getContent`, () => {
    const cache = new Cache(10)
    cache.set(`a`, ast(1))
    cache.set(`b`, ast(2))
    expect(cache.getContent()).toEqual({
      a: ast(1),
      b: ast(2),
    })
  })
  test(`getContent`, () => {
    const cache = new Cache(null)
    cache.set(`a`, ast(1))
    cache.set(`b`, ast(2))
    expect(cache.getContent()).toEqual({
      a: ast(1),
      b: ast(2),
    })
  })

  test(`max cache size must be at least 1`, () => {
    expect(() => new Cache(-1)).toThrow()
    expect(() => new Cache(0)).toThrow()
    expect(() => new Cache(0.1)).not.toThrow()
    expect(() => new Cache(1)).not.toThrow()
  })

  test(`Add an entry.`, () => {
    const cache = new Cache(10)
    expect(cache.size).toBe(0)
    cache.set(`a`, ast(1))
    expect(cache.size).toBe(1)
    expect(cache.get(`a`)).toEqual(ast(1))
    expect(cache.has(`a`)).toBe(true)
  })

  test(`Clear cache.`, () => {
    const cache = new Cache(10)
    cache.set(`a`, ast(1))
    cache.set(`b`, ast(2))
    cache.set(`c`, ast(3))
    expect(cache.size).toBe(3)
    cache.clear()
    expect(cache.size).toBe(0)
  })

  test(`Add an entry - cacheSize = 1`, () => {
    const cache = new Cache(1)
    expect(cache.size).toBe(0)
    cache.set(`a`, ast(1))
    expect(cache.size).toBe(1)
    expect(cache.get(`a`)).toEqual(ast(1))
  })
  test(`maxSize.`, () => {
    const cache = new Cache(1)
    cache.set(`a`, ast(1))
    expect(cache.get(`a`)).toEqual(ast(1))
    cache.set(`b`, ast(2))
    expect(cache.size).toBe(1)
    expect(cache.get(`a`)).toBeUndefined()
    expect(cache.has(`a`)).toBe(false)
    expect(cache.get(`b`)).toEqual(ast(2))
    expect(cache.has(`b`)).toBe(true)
  })
  test(`maxSize.`, () => {
    const cache = new Cache(1)
    cache.set(`a`, ast(1))
    expect(cache.get(`a`)).toEqual(ast(1))
    cache.set(`b`, ast(2))
    expect(cache.size).toBe(1)
    expect(cache.get(`a`)).toBeUndefined()
    expect(cache.has(`a`)).toBe(false)
    expect(cache.get(`b`)).toEqual(ast(2))
    expect(cache.has(`b`)).toBe(true)
  })
})

describe(`regressions`, () => {
  let lits: Lits
  beforeEach(() => {
    lits = new Lits({ debug: true })
  })
  test(`debugInfo`, () => {
    try {
      lits.run(`(loop [n 3]
  (write! n)
  (when (not zero? n))
    (recur (dec n))
  )
)`)
    } catch (error) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      expect((error as any).debugInfo.line).toBe(3)
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      expect((error as any).debugInfo.column).toBe(10)
    }
  })
  test(`name not recognized`, () => {
    expect(() => lits.run(`(asd)`)).toThrowError(UndefinedSymbolError)
    expect(() => lits.run(`asd`)).toThrowError(UndefinedSymbolError)
  })

  test(`debug info when executing function with error in`, () => {
    const program = `(defn formatPhoneNumber [$data]
  (if (string? $data)
    (let [phoneNumber (if (= "+" (nth $data 0)) (subs $data 2) $data)]
      (cond
        (> (count phoneNumber) 6)
          (astr
            "("
            (subs phoneNumber 0 3)

            ") "
            (subs phoneNumber 3 6)
            "-"
            (subs phoneNumber 6))

        (> (count phoneNumber) 3)
          (str "(" (subs phoneNumber 0 3) ") " (subs phoneNumber 3))

        (> (count phoneNumber) 0)
          (str "(" (subs phoneNumber 0))

        true
          phoneNumber
      )
    )
    ""
  )
)

(formatPhoneNumber "+1234232123")`
    try {
      lits.run(program)
      fail()
    } catch (error) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      expect((error as any).debugInfo.line).toBe(6)
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      expect((error as any).debugInfo.column).toBe(12)
    }
  })
  test(`unexpected argument`, () => {
    try {
      lits.run(`(+ 1 + 2)`)
    } catch (error) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const anyError = error as any
      expect(anyError.debugInfo.line).toBe(1)
      expect(anyError.debugInfo.column).toBe(6)
    }
  })
})
