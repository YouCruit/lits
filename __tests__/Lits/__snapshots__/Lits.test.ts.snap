// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runtime info getRuntimeInfo() with ast cache 1`] = `
Object {
  "astCacheSize": 10,
  "currentAstCacheSize": 0,
  "debug": false,
}
`;

exports[`runtime info getRuntimeInfo(). 1`] = `
Object {
  "astCacheSize": 0,
  "currentAstCacheSize": 0,
  "debug": false,
}
`;
